Duke-CIT/edx-platform
# Redirect users to the SSO login

Cherry-picks changes from https://github.com/edx/edx-platform/pull/15587, and backports the following PRs into `ficus.3`, to support the related `third_party_auth` and `python-social-auth` migrations that this change requires.

Note that this change migrates to `python-social-auth==0.2.21`, but does not proceed to the split django auth.  (cd [`python-social-auth` MIGRATING_TO_SOCIAL.md#migrations](https://github.com/omab/python-social-auth/blob/master/MIGRATING_TO_SOCIAL.md#migrations))

- https://github.com/edx/edx-platform/pull/14783: b64503848c1dc50f06f44626c71f93df609ffd12 - 0006_samlproviderconfig_automatic_refresh_enabled
- https://github.com/edx/edx-platform/pull/14793: da3867e870e61ac01b8cf584b01eda6c98939844 - 0007_auto_20170406_0912
- https://github.com/edx/edx-platform/pull/14896: 3f6e690827464c5c542ddc1fbb325c6d6851fdd0 - 0008_auto_20170413_1455
- Upgraded to Python Social Auth 0.2.21 to resolve migration issue: 4cc7628696be0be35dc6eb6af0a496959d263a42
- https://github.com/edx/edx-platform/pull/14900: ee9f632a3dfad988da8788e4cd38d740aa3773e7 - 0009_auto_20170415_1144
- https://github.com/edx/edx-platform/pull/15022 (post social-auth split) 123584e5af3c25d3c608dc1ec6d3c2fd76cba053 - 0010_add_skip_hinted_login_dialog_field
- https://github.com/edx/edx-platform/pull/15587: d5c85331c5b8e9f1042490b47d9bb95eac0d9b6e - 0011_auto_20170616_0112
- https://github.com/edx/edx-platform/pull/15408: ae8a5f2bf7ace2a6993d7c2e333daa09beafae98 - 0012_auto_20170626_1135
- https://github.com/edx/edx-platform/pull/15587 (Open): cc9855ffc7afa8f11e8b23cc7d6e5ea6ba6fd78b

Also backports this change to the Grades API:
- https://github.com/edx/edx-platform/pull/14710: 4d992565ac8428315f4355b5cca78a9225bd6bf4

**JIRA tickets**: [OC-2799](https://tasks.opencraft.com/browse/OC-2799)

**Discussions**: Ginkgo release is not yet ready, so Duke has opted to back-port this change to Ficus.3.

**Screenshots**:

**Sandbox URL**:

Deployed to stage.extend.duke.edu for testing.

**Testing instructions**:

See https://github.com/edx/edx-platform/pull/15587 for manual test instructions.
See also [Integrating with an OAuth2 Identity Provider](http://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/tpa/tpa_integrate_open/tpa_oauth.html) for testing using OAuth2.

```bash
paver test_system --disable-migrations --fasttest -t common/djangoapps/student/tests/test_helpers.py
paver test_system --disable-migrations --fasttest -t common/djangoapps/third_party_auth/management/commands/tests/test_saml.py
paver test_system --disable-migrations --fasttest -t common/djangoapps/third_party_auth/tests/specs/test_testshib.py
paver test_system --disable-migrations --fasttest -t common/djangoapps/third_party_auth/tests/test_admin.py
paver test_system --disable-migrations --fasttest -t common/djangoapps/third_party_auth/tests/test_decorators.py
paver test_system --disable-migrations --fasttest -t lms/djangoapps/student_account/test/test_views.py
paver test_system --disable-migrations --fasttest -t openedx/core/djangoapps/user_api/tests/test_helpers.py
paver test_system --disable-migrations --fasttest -t lms/djangoapps/grades/api/tests/test_views.py
```

**Author notes and concerns**:

1. These tests are failing, but I believe they're due to issues with the cherry-picked tests and the split social auth, not the platform functionality, because both SAML+OAuth2 work during manual testing.
   ```bash
   # ImportError: No module named social_core
   paver test_system --disable-migrations --fasttest -t common/djangoapps/third_party_auth/tests/test_pipeline_integration.py
   # AssertionError: Could not find field social_auth_provider
   paver test_system --disable-migrations --fasttest -t openedx/core/djangoapps/user_api/tests/test_views.py
   ```
   Finalising the migration to split social auth might fix these failures, but it requires merging the changes from https://github.com/edx/edx-platform/pull/15135, which cause a lot of conflicts, and so will take time to do.
  In the interests of saving time, I've opted to hold the migration here, and do this final step when we upgrade to Ginkgo.

**Reviewers**
- [ ] @snowcrshd

**Settings to be added for Duke**
```yaml
EDXAPP_FEATURES:
  THIRD_PARTY_AUTH_HINT: "saml-duke"
```
